src/repository/opamDownload.ml

```diff
@@ -193,3 +193,169 @@ let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
+
+  let get_value key s =
+    match OpamJson.of_string s with
+    | Some (`O elems) ->
+      (match OpamStd.List.assoc_opt key elems with
+       | Some (`String v) -> Some v
+       | _ -> None)
+    | _ -> None
+
+  (* SWH request output example
+     directory: retrieve "status" & "fetch_url"
+     $ curl https://archive.softwareheritage.org/api/1/vault/directory/4453cfbdab1a996658cd1a815711664ee7742380/
+     {
+      "fetch_url": "https://archive.softwareheritage.org/api/1/vault/flat/swh:1:dir:4453cfbdab1a996658cd1a815711664ee7742380/raw/",
+      "progress_message": null,
+      "id": 398307347,
+      "status": "done",
+      "swhid": "swh:1:dir:4453cfbdab1a996658cd1a815711664ee7742380",
+      "obj_type": "directory",
+      "obj_id": "4453cfbdab1a996658cd1a815711664ee7742380"
+     }
+  *)
+
+  let get_output ?(post=false) url =
+    get_output ~post url @@| function
+    | Some out ->
+      Some (String.concat "" out)
+    | None ->
+      OpamConsole.error "Software Heritage fallback need %s or %s installed"
+        (OpamConsole.colorise `underline "curl")
+        (OpamConsole.colorise `underline "wget");
+      None
+
+  let get_dir hash =
+    let url = full_url "vault/directory" hash in
+    get_output ~post:true url @@| OpamStd.Option.replace @@ fun json ->
+    let status = get_value "status" json in
+    let fetch_url = get_value "fetch_url" json in
+    match status, fetch_url with
+    | None, _ | _, None -> None
+    | Some status, Some fetch_url ->
+      Some (match status with
+          | "done" -> `Done (OpamUrl.of_string fetch_url)
+          | "pending" -> `Pending
+          | "new" -> `New
+          | "failed" -> `Failed
+          | _ -> `Unknown)
+
+  let fallback_err fmt = Printf.sprintf ("SWH fallback: "^^fmt)
+
+  let get_url ?(timeout=6) swhid =
+    let attempts = timeout in
+    let hash = OpamSWHID.hash swhid in
+    let rec aux timeout =
+      if timeout <= 0 then
+        Done (Not_available
+                (Some (fallback_err "timeout"),
+                 fallback_err "%d attempts tried, aborting" attempts))
+      else
+        get_dir hash @@+ function
+        | Some (`Done fetch_url) -> Done (Result fetch_url)
+        | Some (`Pending | `New) ->
+          Unix.sleep 10;
+          aux (timeout - 1)
+        | None | Some (`Failed | `Unknown) ->
+          Done (Not_available (None, fallback_err "Unknown swhid"))
+    in
+    aux timeout
+
+  (* for the moment only used in sources, not extra sources or files *)
+  let archive_fallback ?timeout urlf dirnames =
+    match OpamFile.URL.swhid urlf with
+    | None -> Done (Result None)
+    | Some swhid ->
+      if check_post_tool () then
+        (* Add a global modifier and/or command for default answering *)
+        if OpamConsole.confirm ~default:false
+            "Source %s is not available. Do you want to try to retrieve it \
+             from Software Heritage cache? It may take few minutes."
```

Comment:
Possibly include a link to https://www.softwareheritage.org for information about what it is?